hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/main/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicator.java

```diff
@@ -0,0 +1,305 @@
+
+              @Override
+              public Thread newThread(Runnable r) {
+                Thread thread = new Thread(r);
+                thread.setDaemon(true);
+                thread.setName("cloud-bigtable-replication-sink-" + threadCount.incrementAndGet());
+                return thread;
+              }
+            });
+  }
+
+  private static synchronized void getOrCreateBigtableConnection(Configuration configuration) {
+
+    if (numConnectionReference == 0) {
+
+      String projectId = configuration.get(PROJECT_KEY);
+      String instanceId = configuration.get(INSTANCE_KEY);
+      // If an App profile is provided, it will be picked automatically by the connection.
+      connection = BigtableConfiguration.connect(configuration);
+      LOG.info("Created a connection to CBT. " + projectId + "--" + instanceId);
+    }
+
+    numConnectionReference++;
+  }
+
+  public synchronized void start(Configuration configuration, MetricsExporter metricsExporter) {
+    LOG.info("Starting replication to CBT.");
+
+    getOrCreateBigtableConnection(configuration);
+    // Create the executor service for the first time.
+    initExecutorService(configuration);
+    batchSizeThresholdInBytes = configuration.getLong(BATCH_SIZE_KEY, DEFAULT_BATCH_SIZE_IN_BYTES);
+
+    this.incompatibleMutationAdapter =
+        new IncompatibleMutationAdapterFactory(configuration, metricsExporter, connection)
+            .createIncompatibleMutationAdapter();
+  }
+
+  public void stop() {
+
+    LOG.info("Stopping replication to CBT.");
+
+    // Connection is shared by all the instances of this class, close it only if no one is using it.
+    // Closing the connection is required as it owns the underlying gRpc connections and gRpc does
+    // not like JVM shutting without closing the gRpc connections.
+    synchronized (CloudBigtableReplicator.class) {
+      if (--numConnectionReference == 0) {
+        try {
+          LOG.warn("Closing the Bigtable connection.");
+          connection.close();
+        } catch (IOException e) {
+          LOG.error("Failed to close Bigtable connection: ", e);
+        }
+      }
+    }
+  }
+
+  public UUID getPeerUUID() {
+    // UUID is used to de-duplicate mutations and avoid the replication loop. In a fully connected
+    // replication topology, all the HBase clusters should recognize the CBT with same UUID.
+    return UUID.nameUUIDFromBytes("Cloud-bigtable".getBytes(StandardCharsets.UTF_8));
+  }
+
+  /**
+   * Replicates the WAL entries to Cloud Bigtable via HBase client. Returns true if everything is
+   * successfully replicated; returns false if anything fails. On failure, HBase should retry the
+   * whole batch again.
+   *
+   * <p>Does not throw any exceptions to the caller. In case of any errors, should return false.
+   *
+   * @param walEntriesByTable Map of WALEntries keyed by table name.
+   */
+  public boolean replicate(Map<String, List<BigtableWALEntry>> walEntriesByTable) {
```

Comment:
can you extract the per table code into a function, it would make it a lot easier to read:

for(Entry e : walEntriesByTable.entrySet()) {
   futures.add(replicateTable(e.getKey(), e.getValue()))
}
for(Future f : futures) {
  succeeded = f.get() && succeeded;
  ..
}